src/docs/arch-block.dox

```diff
@@ -1,9 +1,192 @@
+created on disk and the associated \c allocation_size is written out to the
+metadata file. However the block manager itself isn't created. Only when a btree
+is opened does WiredTiger allocate a block manager. Each btree has it's own
+block manager.
+
+@subsection read Reading files and pages
+
+When an existing btree is opened for the first time, the location of the root
+block is contained in the metadata file \c WiredTiger.wt. The block manager will
+read the block at the location specified and return the page image as a buffer
+to the layer above. This will then be instantiated as a page in memory.
+
+From there subsequent page addresses can be read from the root page and the
+process repeated as required. If a cursor traverses to a page which hasn't been
+read into memory the same process will take place.
+
+@subsection Writing
+
+Two cases exist for writing out data using the block manager, checkpoint and
+eviction. Both are explained in the following two sections. When a page image
+is written out the block manager the \c bm->write API is called. See \c bt_io.c
+for more detail.
+
+@subsubsection Checkpoint
+
+Checkpoints are the process by which a file can be guaranteed to contain a
+consistent snapshot of the data at a given point in time for more detail on that
+see @ref arch-checkpoint.
+
+At the block manager level a checkpoint corresponds with a set of blocks which
+are stored in the file associated with the given URI. Typically each file will
+contain a minimum of two checkpoints. When opening an existing file the latest
+checkpoint is read.
+
+When WiredTiger performs a checkpoint clean pages will not result in additional
+blocks being created and as such existing blocks from older checkpoints may be
+reused, this means that while there may be more than one checkpoint in a file
+the block are not guaranteed to be mutually exclusive. Checkpoints are created
+in depth first order, leaf blocks are created, then the parent blocks. This is
+a requirement as the parent blocks contain the addresses of the leaf blocks.
+
+The block manager doesn't guarantee that calling \c bm->write will result in
+the data being flushed to disk. In the checkpoint scenario WiredTiger will also
+call \c bm->sync which will call the file system dependent flush function.
+
+As a checkpoint progresses it updates its three extent lists, these extent lists
+are written out to disk as part of the checkpoint in blocks. When a checkpoint
+is written the previous checkpoints will be cleaned up if possible, suppose we
+have finished writing out checkpoint C and have checkpoints A and B on disk.
+We only need to retain two checkpoints so we can cleanup checkpoint A.
+Checkpoint A's extent lists get merged into checkpoint B's extent lists. If
+a block, N, is in the alloc list of checkpoint A and the removed discard list
+of checkpoint B the block is considered free. Until a block in an alloc list
+is seen in a discard list the space cannot freed.
```

Comment:
This describes algorithmically what happens.  I don't know if it would help to provide a higher level intuition of why this works.  I have to work out the details each time I think about this. So it's probably less obvious to folks who haven't thought through it before.  Let me know if you want me to try to propose some text.

Reply:
I should also not use "we" here so I'll try and think of how I can word this a bit better.

Reply:
Agree that this could be worded better. Also happy to propose a change.

Reply:
I tried to re-word this in commit: b4acd91c3 let me know what you think. Personally I think it's okay to be fairly algorithmic here.

Reply:
I'm would definitely include the algorithm description.  But think it would help to add some intuition about why this particular set of extent list operations is sufficient to determine when a block is free. This will also help to show why this works even if there are more than 2 checkpoints, or if I delete a checkpoint from the middle of a sequence of checkpoints. (Or maybe I'm misjudging what is complex vs. obvious here, or useful vs. unimportant) 

Maybe something like:
"Any block will belong to a set of one or more consecutive checkpoints.  In the first of those checkpoints, the block will appear in the alloc list. In the last it will appear in the discard list. Therefore the block manager can use the extent lists to determine which blocks can be freed when deleting a checkpoints, as follows..."

Reply:
I'll kinda separate it it under a heading which will be in bold and i'll work on a more in depth description. I think depth is helpful here as it's a pretty complicated thing so being explicit is the way to go.